client/src/main/java/com/aws/greengrass/cli/adapter/impl/NucleusAdapterIpcClientImpl.java

```diff
@@ -209,6 +225,198 @@ public CreateDebugPasswordResponse createDebugPassword() {
+                    System.err.println("Unauthorized error while subscribing to topic: " + topicName);
+                } else {
+                    throw e;
+                }
+            }
+
+            // Keep the main thread alive, or the process will exit.
+            try {
+                while (true) {
+                    Thread.sleep(10000);
+                }
+            } catch (InterruptedException e) {
+                System.out.println("Subscribe interrupted.");
+            }
+
+            // To stop subscribing, close the stream.
+            responseHandler.closeStream();
+        } catch (InterruptedException e) {
+            System.out.println("IPC interrupted.");
+        } catch (ExecutionException e) {
+            System.err.println("Exception occurred when using IPC.");
+            e.printStackTrace();
+            System.exit(1);
+        }
+    }
+
+    public static class SubscriptionMqttResponseHandler implements StreamResponseHandler<IoTCoreMessage> {
+
+        @Override
+        public void onStreamEvent(IoTCoreMessage ioTCoreMessage) {
+            try {
+                String topic = ioTCoreMessage.getMessage().getTopicName();
+                String message = new String(ioTCoreMessage.getMessage().getPayload(),
+                        StandardCharsets.UTF_8);
+                System.out.printf("Received new message on topic %s: %s%n", topic, message);
+            } catch (Exception e) {
+                System.err.println("Exception occurred while processing subscription response " +
+                        "message.");
+                e.printStackTrace();
+            }
+        }
+
+        @Override
+        public boolean onStreamError(Throwable error) {
+            System.err.println("Received a stream error.");
+            error.printStackTrace();
+            return false;
+        }
+
+        @Override
+        public void onStreamClosed() {
+            System.out.println("Subscribe to IoT Core stream closed.");
+        }
+    }
+
+    private SubscribeToTopicResponseHandler subscribeToTopic(GreengrassCoreIPCClient greengrassCoreIPCClient
+            , String topic, StreamResponseHandler<SubscriptionResponseMessage> streamResponseHandler) {
+        SubscribeToTopicRequest subscribeToTopicRequest = new SubscribeToTopicRequest();
+        subscribeToTopicRequest.setTopic(topic);
+        return greengrassCoreIPCClient.subscribeToTopic(subscribeToTopicRequest,
+                Optional.of(streamResponseHandler));
+    }
+
+
+    public static class SubscriptionResponseHandler implements StreamResponseHandler<SubscriptionResponseMessage> {
+
+        private final String topic;
+
+        public SubscriptionResponseHandler(String topic) {
+            this.topic = topic;
+        }
+
+        @Override
+        public void onStreamEvent(SubscriptionResponseMessage subscriptionResponseMessage) {
+            try {
+                String message =
```

Comment:
handle the case where the message is a jsonmessage type. Print the json message as a json string